### ISSUE TITLE
StCR Import Bug: Blacklisted Email Addresses prevent certain subscriptions from being imported

### ISSUE THREAD

**Issue by raamdev** [websharks/comment-mail#179]
While researching https://github.com/websharks/comment-mail/issues/166, I discovered that the Comment Mail → Config. Options → Blacklisted Email Addresses (which has a large default list of blacklisted email accounts), affects the StCR import process. 

**If a StCR subscription email address matches a Blacklisted Email Address, the import silently excludes that subscription from the import!**

This is a bug because we're _importing existing subscriptions_, i.e., we should assume that if there's an existing active StCR subscription in the database, the site owner expects that subscription to be imported.

---

On this same topic: I suggest getting rid of the default blacklisted email addresses altogether. Having such a large list of default blacklisted email addresses, _without the site owner realizing many addresses are already blacklisted_, is going to lead to confusion. For example, I see `admin@*` is on the default Blacklist...I've seen _many_ valid email addresses that use `admin@`; in fact I even use one myself!

IMO, it would be better to allow the site owner to build their own blacklist as they see fit, instead of assuming a default list.

**Comment by jaswsinc** [comment]
I agree. Although, I think it would be nice to have an option during import. Role-based emails like `admin@` are normally filtered by default in most mass-email systems, because it protects the integrity of whatever mail server is in use; e.g., http://blog.mailchimp.com/role-addresses-are-not-people/

So for instance, while you might feel it's a bug, another person might see this as a feature. So a checkbox next to the import button might be helpful in that respect. Giving the site owner a choice.

I think the default list of blacklisted role-based emails should be improved (i.e., made longer), because that's a nice library of role-based addresses that you really shouldn't be sending email to. Needing to look those up and build the list yourself would be harder than it coming with CM by default.

**Comment by raamdev** [comment]
If it's made an option and disabled by default, that's fine with me. :-)

**Comment by raamdev** [comment]
Now that the StCR Importer has been updated to track number of skipped StCR subscriptions (see #166), including those skipped due to a Blacklisted Email Address, I'm changing this GitHub Issue to an `enhancement` so that we can work towards adding an option to have the StCR Importer ignore the blacklist.

My feeling is still that this option should be enabled by default (i.e., _don't_ apply the Blacklisted Email Addresses to imported subscriptions by default).

**Comment by jaswsinc** [comment]
## Next Actions (Step 1)

- Add a new option key here. After [this line](https://github.com/websharks/comment-mail-pro/blob/151224/comment-mail-pro/includes/classes/sub-inserter.php#L359) add the following:

  ```php
  'check_blacklist' => true,
  ```

- Declare that new option key as a new class property also. Duplicate [this block](https://github.com/websharks/comment-mail-pro/blob/151224/comment-mail-pro/includes/classes/sub-inserter.php#L157-L162) and add the new class property declaration for `protected $check_blacklist;`

- Also duplicate [this line](https://github.com/websharks/comment-mail-pro/blob/151224/comment-mail-pro/includes/classes/sub-inserter.php#L399) in order to set the property value for `$this->check_blacklist` in the same way we set `$this->keep_existing`.

- Now, in [this line](https://github.com/websharks/comment-mail-pro/blob/151224/comment-mail-pro/includes/classes/sub-inserter.php#L1255) we should only check the blacklist if `$this->check_blacklist` is true.

**Comment by jaswsinc** [comment]
## On Deck...

Once the `sub_inserter{}` class is updated (Step 1 above) to allow for `'check_blacklist' => false`, we can search for calls to the `sub_inserter{}` class in [this file](https://github.com/websharks/comment-mail-pro/blob/000000-dev/comment-mail-pro/includes/classes/import-stcr.php) and falsify that flag by default.

_Then work to make it a configurable option perhaps._

**Comment by kristineds** [comment]
@jaswsinc @raamdev Submitting PRs for Step 1.

- [PR for Pro](https://github.com/websharks/comment-mail-pro/pull/44)
- [PR for Lite](https://github.com/websharks/comment-mail/pull/198)

> search for calls to the sub_inserter{} class in this file and falsify that flag by default.

How do we falsify that flag by default and search for the calls? Can you give an example? :)

**Comment by jaswsinc** [comment]
@kristineds writes...

> How do we falsify that flag by default and search for the calls? Can you give an example? :)

You can look at the class constructor for the `sub_inserter{}` here.
https://github.com/websharks/comment-mail-pro/blob/151224/comment-mail-pro/includes/classes/sub-inserter.php#L286

Notice that it accepts two arguments, where each of those arguments are an array. The first array is the request data; i.e., subscriber data that is to be inserted. The second argument can be an array of behavioral details.

---

In the current StCR importer, we only pass the request args. No behavior arguments. However, we could start doing this in order to falsify that specific flag which disables checks against the blacklist now.

See: https://github.com/websharks/comment-mail-pro/blob/151224/comment-mail-pro/includes/classes/import-stcr.php#L250-L262

Example:

```php
$_sub_inserter = new sub_inserter($_sub_insert_data, array('check_blacklist' => false));
```

**Comment by jaswsinc** [comment]
@kristineds Just a reminder that the two open PRs for this issue still have a minor problem that needs to be fixed so we can merge. See: https://github.com/websharks/comment-mail-pro/pull/44#commitcomment-15337838

Also, if you can please leave me your thoughts about [my comment here](https://github.com/websharks/comment-mail/issues/179#issuecomment-169955438), and let me know if you need help with that or not. If you do, I can build a new outline for the second portion of this work.

**Comment by kristineds** [comment]
@jaswsinc writes..
> Just a reminder that the two open PRs for this issue still have a minor problem that needs to be fixed so we can merge. See: websharks/comment-mail-pro#44 (comment)

Here are the updated PRs to fix those minor problems:
- [PR for PRO](https://github.com/websharks/comment-mail-pro/commit/0d9716275dd48bc9f5a72082d7834f43dfc64996)
- [PR for Lite](https://github.com/websharks/comment-mail/commit/4605a463aa9b54eb1ef91b64dd7d317d992fdca7)

> Also, if you can please leave me your thoughts about my comment here, and let me know if you need help with that or not. If you do, I can build a new outline for the second portion of this work.

Still trying to wrap my head around this. If you could build the outline, I might understand it more. Thank you! :)

**Comment by raamdev** [comment]
@kristineds @jaswsinc I've merged PRs https://github.com/websharks/comment-mail/pull/198 and https://github.com/websharks/comment-mail-pro/pull/44 so work on this specific feature request (as it relates to the StCR importer) can continue in a separate PR.
